x-pack/plugins/enterprise_search/public/applications/enterprise_search_content/components/search_index/crawler/crawler_domain_detail/extraction_rules/edit_extraction_rule.tsx

```diff
@@ -0,0 +1,442 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useEffect, useState } from 'react';
+
+import { Controller, useFieldArray, useForm } from 'react-hook-form';
+
+import { useActions, useValues } from 'kea';
+
+import {
+  EuiButton,
+  EuiButtonEmpty,
+  EuiButtonIcon,
+  EuiFieldText,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiForm,
+  EuiFormRow,
+  EuiLink,
+  EuiPanel,
+  EuiRadioGroup,
+  EuiSelect,
+  EuiSpacer,
+  EuiTitle,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+
+import {
+  ExtractionFilter,
+  ExtractionRule,
+  ExtractionRuleBase,
+} from '../../../../../../../../common/types/extraction_rules';
+
+import { ContentFieldsPanel } from './content_fields_panel';
+import { EditFieldRuleFlyout } from './edit_field_rule_flyout';
+import { ExtractionRulesLogic } from './extraction_rules_logic';
+
+interface EditExtractionRuleProps {
+  cancelEditing: () => void;
+  extractionRule: ExtractionRule | null;
+  isNewRule: boolean;
+  saveRule: (rule: ExtractionRuleBase) => void;
+}
+
+enum UrlState {
+  ALL = 'all',
+  SPECIFIC = 'specific',
+}
+
+const getReadableExtractionFilter = (rule: ExtractionFilter) => {
+  switch (rule) {
+    case ExtractionFilter.BEGINS:
+      return i18n.translate(
+        'xpack.enterpriseSearch.crawler.extractionRulesExtractionFilter.beginsWithLabel',
+        {
+          defaultMessage: 'Begins with',
+        }
+      );
+    case ExtractionFilter.ENDS:
+      return i18n.translate(
+        'xpack.enterpriseSearch.crawler.extractionRulesExtractionFilter.endsWithLabel',
+        {
+          defaultMessage: 'Ends with',
+        }
+      );
+    case ExtractionFilter.CONTAINS:
+      return i18n.translate(
+        'xpack.enterpriseSearch.crawler.extractionRulesExtractionFilter.containsLabel',
+        {
+          defaultMessage: 'Contains',
+        }
+      );
+    case ExtractionFilter.REGEX:
+      return i18n.translate(
+        'xpack.enterpriseSearch.crawler.extractionRulesExtractionFilter.regexLabel',
+        {
+          defaultMessage: 'Regex',
+        }
+      );
+  }
+};
+
+const extractionFilterOptions = [
+  ExtractionFilter.BEGINS,
+  ExtractionFilter.ENDS,
+  ExtractionFilter.CONTAINS,
+  ExtractionFilter.REGEX,
+].map((ruleOption: ExtractionFilter) => ({
+  text: getReadableExtractionFilter(ruleOption),
+  value: ruleOption,
+}));
+
+export const EditExtractionRule: React.FC<EditExtractionRuleProps> = ({
+  cancelEditing,
+  extractionRule,
+  isNewRule,
+  saveRule,
+}) => {
+  const { closeEditRuleFlyout, openEditRuleFlyout } = useActions(ExtractionRulesLogic);
+  const { fieldRuleFlyoutVisible, fieldRuleToEdit, fieldRuleToEditIndex, fieldRuleToEditIsNew } =
+    useValues(ExtractionRulesLogic);
+  const [urlToggle, setUrlToggle] = useState<UrlState>(UrlState.ALL);
+  const { control, formState, getValues, handleSubmit, reset, setValue } =
+    useForm<ExtractionRuleBase>({
```

Comment:
nice :) useFieldArray work nicely here.

Reply:
Yeah it was perfect for this use case. I'm sure I did some things suboptimally, though, and it took a little time to figure out how to best make use react form interact nicely with our EUI components.